src/ir/can_derive_debug.rs

```diff
@@ -0,0 +1,271 @@
+            EdgeKind::InnerType |
+            EdgeKind::InnerVar => true,
+
+            EdgeKind::Constructor |
+            EdgeKind::FunctionReturn |
+            EdgeKind::FunctionParameter |
+            EdgeKind::Method => false,
+            EdgeKind::Generic => false,
+        }
+    }
+}
+
+impl<'ctx, 'gen> MonotoneFramework for CanDeriveDebugAnalysis<'ctx, 'gen> {
+    type Node = ItemId;
+    type Extra = &'ctx BindgenContext<'gen>;
+    type Output = HashSet<ItemId>;
+
+    fn new(ctx: &'ctx BindgenContext<'gen>) -> CanDeriveDebugAnalysis<'ctx, 'gen> {
+        let can_derive_debug = HashSet::new();
+        let mut dependencies = HashMap::new();
+        let whitelisted_items: HashSet<_> = ctx.whitelisted_items().collect();
+
+        let whitelisted_and_blacklisted_items: ItemSet = whitelisted_items.iter()
+            .cloned()
+            .flat_map(|i| {
+                let mut reachable = vec![i];
+                i.trace(ctx, &mut |s, _| {
+                    reachable.push(s);
+                }, &());
+                reachable
+            })
+            .collect();
+
+        for item in whitelisted_and_blacklisted_items {
+            dependencies.entry(item).or_insert(vec![]);
+
+            {
+                // We reverse our natural IR graph edges to find dependencies
+                // between nodes.
+                item.trace(ctx, &mut |sub_item: ItemId, edge_kind| {
+                    if Self::consider_edge(edge_kind) {
+                        dependencies.entry(sub_item)
+                            .or_insert(vec![])
+                            .push(item);
+                    }
+                }, &());
+            }
+        }
+
+        CanDeriveDebugAnalysis {
+            ctx: ctx,
+            can_derive_debug: can_derive_debug,
+            dependencies: dependencies,
+        }
+    }
+
+    fn initial_worklist(&self) -> Vec<ItemId> {
+        self.ctx.whitelisted_items().collect()
+    }
+
+    fn constrain(&mut self, id: ItemId) -> bool {
+        if self.can_derive_debug.contains(&id) {
+            return false;
+        }
+
+        let item = self.ctx.resolve_item(id);
+        let ty = match item.as_type() {
+            None => return false,
+            Some(ty) => ty
+        };
+
+        match ty.kind() {
+            &TypeKind::Void |
+            &TypeKind::NullPtr |
+            &TypeKind::Int(..) |
+            &TypeKind::Float(..) |
+            &TypeKind::Complex(..) |
+            &TypeKind::Function(..) |
+            &TypeKind::Enum(..) |
+            &TypeKind::Reference(..) |
+            &TypeKind::BlockPointer |
+            &TypeKind::Named |
+            &TypeKind::UnresolvedTypeRef(..) |
+            &TypeKind::ObjCInterface(..) |
+            &TypeKind::ObjCId |
+            &TypeKind::ObjCSel => {
+                self.can_derive_debug.insert(id);
+                true
+            },
+            &TypeKind::Opaque => {
+                if ty.layout(self.ctx)
```

Comment:
Let's try and translate exactly what master is doing into the constraint function.